lib/node/NodeMainTemplatePlugin.js

```diff
@@ -2,130 +2,160 @@
-					"});"
-				]),
-				"};"
-			]);
-		}
-		return source;
-	});
-	mainTemplate.plugin("require-ensure", function(_, chunk, hash) {
-		var chunkFilename = this.outputOptions.chunkFilename;
-		var chunkMaps = chunk.getChunkMaps();
-		var insertMoreModules = [
-			"var moreModules = chunk.modules, chunkIds = chunk.ids;",
-			"for(var moduleId in moreModules) {",
-			this.indent(this.renderAddModule(hash, chunk, "moduleId", "moreModules[moduleId]")),
-			"}"
-		];
-		if(self.asyncChunkLoading) {
-			return this.asString([
-				"// \"0\" is the signal for \"already loaded\"",
-				"if(installedChunks[chunkId] === 0)",
-				this.indent([
-					"return Promise.resolve();"
-				]),
-				"// array of [resolve, reject, promise] means \"currently loading\"",
-				"if(installedChunks[chunkId])",
-				this.indent([
-					"return installedChunks[chunkId][2];"
-				]),
-				"// load the chunk and return promise to it",
-				"var promise = new Promise(function(resolve, reject) {",
-				this.indent([
-					"installedChunks[chunkId] = [resolve, reject];",
-					"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify("/" + chunkFilename), {
-						hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
-						hashWithLength: function(length) {
-							return "\" + " + this.renderCurrentHashCode(hash, length) + " + \"";
-						}.bind(this),
-						chunk: {
-							id: "\" + chunkId + \"",
-							hash: "\" + " + JSON.stringify(chunkMaps.hash) + "[chunkId] + \"",
-							hashWithLength: function(length) {
-								var shortChunkHashMap = {};
-								Object.keys(chunkMaps.hash).forEach(function(chunkId) {
-									if(typeof chunkMaps.hash[chunkId] === "string")
-										shortChunkHashMap[chunkId] = chunkMaps.hash[chunkId].substr(0, length);
-								});
-								return "\" + " + JSON.stringify(shortChunkHashMap) + "[chunkId] + \"";
-							},
-							name: "\" + (" + JSON.stringify(chunkMaps.name) + "[chunkId]||chunkId) + \""
-						}
-					}) + ";",
-					"require('fs').readFile(filename, 'utf-8',  function(err, content) {",
+const Template = require("../Template");
+
+module.exports = class NodeMainTemplatePlugin {
+	constructor(asyncChunkLoading) {
+		this.asyncChunkLoading = asyncChunkLoading;
+	}
+
+	apply(mainTemplate) {
+		const asyncChunkLoading = this.asyncChunkLoading;
+		mainTemplate.plugin("local-vars", function(source, chunk) {
+			if(chunk.chunks.length > 0) {
+				return this.asString([
+					source,
+					"",
+					"// object to store loaded chunks",
+					"// \"0\" means \"already loaded\"",
+					"var installedChunks = {",
+					this.indent(chunk.ids.map((id) => id + ": 0").join(",\n")),
+					"};"
+				]);
+			}
+			return source;
+		});
+		mainTemplate.plugin("require-extensions", function(source, chunk) {
+			if(chunk.chunks.length > 0) {
+				return this.asString([
+					source,
+					"",
+					"// uncatched error handler for webpack runtime",
+					this.requireFn + ".oe = function(err) {",
```

Comment:
I've thought about it, but then I found https://jsperf.com/es6-string-literals-vs-string-concatenation/14 and seems like string concatenation is faster than template literals, but template literal is faster than array.join (which is used in Template.asString).
Looking on the code in other modules I couldn't find the prefered way to deal with concatenation - in some places it's `+` in other places it's template literals.

Do you prefer template literals?